File: entity-store/src/main/java/jetbrains/exodus/entitystore/EntityIterableCacheAdapter.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.entitystore;

import jetbrains.exodus.core.dataStructures.hash.HashMap;
import jetbrains.exodus.core.dataStructures.hash.ObjectProcedure;
import jetbrains.exodus.core.dataStructures.persistent.EvictListener;
import jetbrains.exodus.core.dataStructures.persistent.PersistentObjectCache;
import jetbrains.exodus.core.execution.SharedTimer;
import jetbrains.exodus.entitystore.iterate.CachedInstanceIterable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.ref.SoftReference;

class EntityIterableCacheAdapter {

    @NotNull
    protected final PersistentEntityStoreConfig config;
    @NotNull
    protected final NonAdjustablePersistentObjectCache<EntityIterableHandle, CacheItem> cache;
    @NotNull
    protected final HashMap<EntityIterableHandle, Updatable> stickyObjects;

    EntityIterableCacheAdapter(@NotNull final PersistentEntityStoreConfig config) {
        this(config, new NonAdjustablePersistentObjectCache<EntityIterableHandle, CacheItem>(config.getEntityIterableCacheSize()), new HashMap<EntityIterableHandle, Updatable>());
    }

    EntityIterableCacheAdapter(@NotNull final PersistentEntityStoreConfig config,
                               @NotNull final NonAdjustablePersistentObjectCache<EntityIterableHandle, CacheItem> cache,
                               @NotNull final HashMap<EntityIterableHandle, Updatable> stickyObjects) {
        this.config = config;
        this.cache = cache;
        this.stickyObjects = stickyObjects;
    }

    @NotNull
    public PersistentObjectCache<EntityIterableHandle, CacheItem> getCacheInstance() {
        return cache;
    }

    @Nullable
    CachedInstanceIterable tryKey(@NotNull final EntityIterableHandle key) {
        if (key.isSticky()) {
            return (CachedInstanceIterable) getStickyObject(key);
        }
        return parseCachedObject(key, cache.tryKey(key));
    }

    @Nullable
    CachedInstanceIterable getObject(@NotNull final EntityIterableHandle key) {
        if (key.isSticky()) {
            return (CachedInstanceIterable) getStickyObject(key);
        }
        return parseCachedObject(key, cache.getObject(key));
    }

    @Nullable
    Updatable getUpdatable(@NotNull final EntityIterableHandle key) {
        if (key.isSticky()) {
            return getStickyObject(key);
        }
        return (Updatable) parseCachedObject(key, cache.getObject(key));
    }

    void cacheObject(@NotNull final EntityIterableHandle key, @NotNull final CachedInstanceIterable it) {
        cache.cacheObject(key, new CacheItem(it, config.getEntityIterableCacheMaxSizeOfDirectValue()));
    }

    void forEachKey(final ObjectProcedure<EntityIterableHandle> procedure) {
        cache.forEachKey(procedure);
    }

    void remove(@NotNull final EntityIterableHandle key) {
        cache.remove(key);
    }

    float hitRate() {
        return cache.hitRate();
    }

    int count() {
        return cache.count();
    }

    int size() {
        return cache.size();
    }

    void clear() {
        cache.clear();
    }

    boolean isSparse() {
        return cache.count() < cache.size() / 2;
    }

    EntityIterableCacheAdapterMutable getClone() {
        return EntityIterableCacheAdapterMutable.create(this);
    }

    void adjustHitRate() {
        cache.adjustHitRate();
    }

    Updatable getStickyObject(@NotNull final EntityIterableHandle handle) {
        Updatable result = stickyObjects.get(handle);
        if (result == null) {
            throw new IllegalStateException("Sticky object not found");
        }
        return result;
    }

    private CachedInstanceIterable parseCachedObject(@NotNull final EntityIterableHandle key, @Nullable final CacheItem item) {
        if (item == null) {
            return null;
        }
        CachedInstanceIterable cached = item.cached;
        if (cached == null) {
            cached = item.ref.get();
            if (cached == null) {
                cache.remove(key);
            }
        }
        return cached;
    }

    static CachedInstanceIterable getCachedValue(@NotNull CacheItem item) {
        CachedInstanceIterable cached = item.cached;
        if (cached == null) {
            cached = item.ref.get();
        }
        return cached;
    }

    static final class CacheItem {
        private final CachedInstanceIterable cached;
        private final SoftReference<CachedInstanceIterable> ref;

        private CacheItem(@NotNull final CachedInstanceIterable it, final int maxSizeOfDirectValue) {
            if (it.isUpdatable() || it.size() <= maxSizeOfDirectValue) {
                cached = it;
                ref = null;
            } else {
                cached = null;
                ref = new SoftReference<>(it);
            }
        }
    }

    /*
    NonAdjustablePersistentObjectCache doesn't adjust itself in order to avoid as many cache adjusters
    as many versions of the cache (as a persistent data structure) can be.
     */
    static class NonAdjustablePersistentObjectCache<K, V> extends PersistentObjectCache<K, V> {

        private NonAdjustablePersistentObjectCache(final int size) {
            super(size);
        }

        private NonAdjustablePersistentObjectCache(@NotNull final NonAdjustablePersistentObjectCache<K, V> source,
                                                   @Nullable final EvictListener<K, V> listener) {
            super(source, listener);
        }

        @Override
        public NonAdjustablePersistentObjectCache<K, V> getClone(final EvictListener<K, V> listener) {
            return new NonAdjustablePersistentObjectCache<>(this, listener);
        }

        NonAdjustablePersistentObjectCache<K, V> endWrite() {
            return new NonAdjustablePersistentObjectCache<>(this, null);
        }

        @Nullable
        @Override
        protected SharedTimer.ExpirablePeriodicTask getCacheAdjuster() {
            return null;
        }
    }
}

File: entity-store/src/main/kotlin/jetbrains/exodus/entitystore/EntityIterableCacheAdapter.kt
/*
 * Copyright ${inceptionYear} - ${year} ${owner}
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.entitystore

import jetbrains.exodus.core.cache.persistent.*
import jetbrains.exodus.entitystore.iterate.CachedInstanceIterable
import mu.KLogging
import java.time.Duration

internal open class EntityIterableCacheAdapter(
    val config: PersistentEntityStoreConfig,
    val cache: PersistentCache<EntityIterableHandle, CachedInstanceIterable>,
    val stickyObjects: HashMap<EntityIterableHandle, Updatable>,
    val cacheKeyIndex: EntityIterableCacheKeyIndex,
) {

    companion object : KLogging() {

        fun create(config: PersistentEntityStoreConfig): EntityIterableCacheAdapter {
            val sizeEviction = if (config.entityIterableCacheSize > 0) {
                val maxSize = config.entityIterableCacheSize.toLong()
                logger.info("Using sized eviction for entity iterable cache, maxSize = $maxSize")
                SizedEviction<CachedInstanceIterable>(config.entityIterableCacheSize.toLong())
            } else {
                val maxWeight = config.entityIterableCacheWeight
                logger.info("Using weighted eviction for entity iterable cache, maxWeight = $maxWeight")
                WeightedEviction(config.entityIterableCacheWeight) { it.roughSize.toInt() }
            }
            val cacheConfig = CaffeineCacheConfig(
                sizeEviction = sizeEviction,
                expireAfterAccess = Duration.ofSeconds(config.entityIterableCacheExpireAfterAccess.toLong()),
                useSoftValues = config.entityIterableCacheSoftValues,
            )

            val index = EntityIterableCacheKeyIndex()
            val cache = CaffeinePersistentCache.create<EntityIterableHandle, CachedInstanceIterable>(cacheConfig, index)

            return EntityIterableCacheAdapter(config, cache, HashMap(), index)
        }
    }

    open fun getObject(key: EntityIterableHandle): CachedInstanceIterable? {
        if (key.isSticky) {
            return getStickyObject(key) as CachedInstanceIterable
        }
        return getStickyObjectUnsafe(key) as CachedInstanceIterable? ?: cache.get(key)
    }

    open fun getUpdatable(key: EntityIterableHandle): Updatable? {
        if (key.isSticky) {
            return getStickyObject(key)
        }
        return getStickyObjectUnsafe(key) ?: cache.get(key) as Updatable?
    }

    open fun cacheObject(key: EntityIterableHandle, it: CachedInstanceIterable) {
        // if it is Updatable then it could be mutated in a txn being stored as sticky object
        if (it is Updatable && stickyObjects.containsKey(key)) {
            stickyObjects[key] = it
        } else {
            cache.put(key, it)
        }
    }

    open fun remove(key: EntityIterableHandle) {
        cache.remove(key)
    }

    open fun count() = cache.count()

    fun size() = cache.size()

    open fun clear() = cache.clear()

    fun registerClient(): PersistentCacheClient {
        return cache.registerClient()
    }

    fun release() {
        cache.release()
    }

    val halfFull: Boolean get() = cache.count() > cache.size() / 2

    fun cloneToMutable(): EntityIterableCacheAdapterMutable {
        return EntityIterableCacheAdapterMutable.cloneFrom(this)
    }

    fun getStickyObjectUnsafe(handle: EntityIterableHandle): Updatable? {
        return stickyObjects[handle]
    }

    fun getStickyObject(handle: EntityIterableHandle): Updatable {
        return getStickyObjectUnsafe(handle) ?: throw IllegalStateException(
            "Sticky object not found, handle=${EntityIterableCache.toString(config, handle)}"
        )
    }
}